File: 2019/lib/day1.ml
(*******************************************************************************
 *                               _            __
 *                              | |          /_ |
 *                            __| | __ _ _   _| |
 *                           / _` |/ _` | | | | |
 *                          | (_| | (_| | |_| | |
 *                           \__,_|\__,_|\__, |_|
 *                                        __/ |
 *                                       |___/
 *
 * Jonathan Lowe
 * github : https://github.com/jglowe
 *
 * The file for day1 of advent of code 2019
 ******************************************************************************)

let calculate_fuel mass = if (mass / 3) - 2 < 0 then 0 else (mass / 3) - 2

let calculate_recursive_fuel mass =
  let rec calculate_recursive_fuel total_fuel mass =
    let fuel = calculate_fuel mass in
    if fuel > 0 then calculate_recursive_fuel (total_fuel + fuel) fuel
    else total_fuel
  in
  calculate_recursive_fuel 0 mass

let part1 strings =
  let masses = List.map int_of_string strings in
  let fuels = List.map calculate_fuel masses in
  List.fold_left ( + ) 0 fuels

let part2 strings =
  let masses = List.map int_of_string strings in
  let fuels = List.map calculate_recursive_fuel masses in
  List.fold_left ( + ) 0 fuels

let%test "Mass Calculation 1" = calculate_fuel 12 = 2

let%test "Mass Calculation 2" = calculate_fuel 14 = 2

let%test "Mass Calculation 3" = calculate_fuel 1969 = 654

let%test "Mass Calculation 4" = calculate_fuel 100756 = 33583

let%test "part1" = part1 ["12"; "14"; "1969"; "100756"] = 34241

let%test "Mass Calculation 1" = calculate_recursive_fuel 12 = 2

let%test "Mass Calculation 2" = calculate_recursive_fuel 14 = 2

let%test "Mass Calculation 3" = calculate_recursive_fuel 1969 = 966

let%test "Mass Calculation 4" = calculate_recursive_fuel 100756 = 50346

let%test "part2" = part1 ["12"; "14"; "1969"; "100756"] = 34241

File: 2023/src/bin/day1.rs
///////////////////////////////////////////////////////////////////////////////////////////////////
//
//                                           _            __
//                                          | |          /_ |
//                                        __| | __ _ _   _| |
//                                       / _` |/ _` | | | | |
//                                      | (_| | (_| | |_| | |
//                                       \__,_|\__,_|\__, |_|
//                                                    __/ |
//                                                   |___/
//
// Jonathan Lowe
// github : https://github.com/jglowe
//
// The file for day1 advent of code 2023
///////////////////////////////////////////////////////////////////////////////////////////////////

fn first_and_last_digits(line: &str) -> u32 {
    let mut first_digit: Option<u32> = None;
    let mut last_digit: Option<u32> = None;

    for item in line.chars() {
        if item.is_digit(10) {
            match (first_digit, last_digit) {
                (None, None) => first_digit = Some(item.to_digit(10).unwrap()),
                (Some(_), None) => last_digit = Some(item.to_digit(10).unwrap()),
                (Some(_), Some(_)) => last_digit = Some(item.to_digit(10).unwrap()),
                (None, Some(_)) => panic!("First digit must be populated first"),
            }
        }
    }

    match (first_digit, last_digit) {
        (Some(x), Some(y)) => x * 10 + y,
        (Some(x), None) => x * 10 + x,
        _ => panic!("Something went wrong, please consult this error in the code"),
    }
}

fn first_and_last_digits_also_spelled_out(line: &str) -> u32 {
    let mut first_digit: Option<u32> = None;
    let mut last_digit: Option<u32> = None;

    for (index, item) in line.chars().enumerate() {
        if item.is_digit(10) {
            match (first_digit, last_digit) {
                (None, None) => first_digit = Some(item.to_digit(10).unwrap()),
                (Some(_), None) => last_digit = Some(item.to_digit(10).unwrap()),
                (Some(_), Some(_)) => last_digit = Some(item.to_digit(10).unwrap()),
                (None, Some(_)) => panic!("First digit must be populated first"),
            }
        } else {
            let mut detected_digit: Option<u32> = None;
            if line.len() >= index + 3 && "one".eq(&line[index..index + 3]) {
                detected_digit = Some(1);
            }
            if line.len() >= index + 3 && "two".eq(&line[index..index + 3]) {
                detected_digit = Some(2);
            }
            if line.len() >= index + 5 && "three".eq(&line[index..index + 5]) {
                detected_digit = Some(3);
            }
            if line.len() >= index + 4 && "four".eq(&line[index..index + 4]) {
                detected_digit = Some(4);
            }
            if line.len() >= index + 4 && "five".eq(&line[index..index + 4]) {
                detected_digit = Some(5);
            }
            if line.len() >= index + 3 && "six".eq(&line[index..index + 3]) {
                detected_digit = Some(6);
            }
            if line.len() >= index + 5 && "seven".eq(&line[index..index + 5]) {
                detected_digit = Some(7);
            }
            if line.len() >= index + 5 && "eight".eq(&line[index..index + 5]) {
                detected_digit = Some(8);
            }
            if line.len() >= index + 4 && "nine".eq(&line[index..index + 4]) {
                detected_digit = Some(9);
            }
            match detected_digit {
                Some(detected_digit) => {
                    match (first_digit, last_digit) {
                        (None, None) => first_digit = Some(detected_digit),
                        (Some(_), None) => last_digit = Some(detected_digit),
                        (Some(_), Some(_)) => last_digit = Some(detected_digit),
                        (None, Some(_)) => panic!("First digit must be populated first"),
                    };
                }
                None => {}
            }
        }
    }

    match (first_digit, last_digit) {
        (Some(x), Some(y)) => x * 10 + y,
        (Some(x), None) => x * 10 + x,
        _ => panic!("Something went wrong, please consult this error in the code"),
    }
}

fn part1(input: &str) -> u32 {
    input.trim().lines().map(first_and_last_digits).sum()
}

fn part2(input: &str) -> u32 {
    input
        .trim()
        .lines()
        .map(first_and_last_digits_also_spelled_out)
        .sum()
}

fn main() {
    let input = advent::read_file("data/day1_input.txt");
    println!("PART 1: {}", part1(&input));
    println!("PART 2: {}", part2(&input));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn part1_test() {
        let input = "1abc2
pqr3stu8vwx
a1b2c3d4e5f
treb7uchet";

        assert_eq!(142, part1(&input));
    }

    #[test]
    fn part2_test() {
        let input = "two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen";

        assert_eq!(281, part2(&input));
    }
}
